plugins/circleci/src/circleci-config.ts

```diff
@@ -1,29 +1,178 @@
-import { promises as fs } from 'fs'
-import merge from 'lodash/merge'
-import path from 'path'
-import YAML from 'yaml'
 import { ToolKitError } from '@dotcom-tool-kit/error'
 import { styles } from '@dotcom-tool-kit/logger'
 import { getOptions } from '@dotcom-tool-kit/options'
 import { Hook } from '@dotcom-tool-kit/types'
+import { automatedComment, CircleConfig, Job, JobConfig, Workflow } from '@dotcom-tool-kit/types/lib/circleci'
 import { semVerRegex } from '@dotcom-tool-kit/types/lib/npm'
-import { Workflow, JobConfig, CircleConfig, automatedComment, Job } from '@dotcom-tool-kit/types/lib/circleci'
+import type { CircleCIOptions } from '@dotcom-tool-kit/types/lib/schema/circleci'
+import { promises as fs } from 'fs'
+import isMatch from 'lodash/isMatch'
+import merge from 'lodash/merge'
+import mergeWith from 'lodash/mergeWith'
+import path from 'path'
+import type { PartialDeep } from 'type-fest'
+import YAML from 'yaml'
+
+const majorOrbVersion = '3'
 
-const majorOrbVersion = '2'
+export type CircleCIState = CircleConfig
+/**
+ * CircleCIStatePartial makes every property in the config object optional,
+ * including nested properties. This is particularly useful as we merge each
+ * hook's config into the larger state, and so each hook only needs to define
+ * the parts of the config they want to add to.
+ */
```

Comment:
**praise:** I love how well documented this code is, I think it's really helpful with something as complex as this Thank you! 🙌